### ISSUE TITLE
Error `could not find parent resource` on a GCP serviceAccount during `pulumi up --target` after updating from v6 to v7

### ISSUE THREAD

**Issue by zbuchheit** [pulumi/pulumi#15889]
### What happened?

>Note: This might actually be more of an engine or pulumi/pulumi issue, it is particularly relevant with the provider name change in gcp `serviceAccount` to `serviceaccount`

While attempting to upgrade from major version v6 to major version v7, I would like to be able to target just the serviceaccount name change so as to prevent running a fully `pulumi up` and carefully check my resources as I do a major update. 

When attempting to use `--target` any resource but the serviceaccount, I will encounter the following error
`error: could not find parent resource urn:pulumi:dev::service-account::gcp:serviceAccount/account:Account::service-account`

If I target the serviceaccount directly it will show a no-op and update the name in the state and going forward everything works correctly and as expected.

### Example

<details>
<summary>
Minified Code Repro
</summary>

```yaml
name: service-account
runtime: yaml
description: A minimal Google Cloud Pulumi YAML program
resources:
  my-bucket-zbuchheit:
    properties:
      location: US
    type: gcp:storage:Bucket
  service-account:
    type: gcp:serviceAccount:Account
    # type: gcp:serviceaccount:Account
    properties:
      accountId: zbuchheit-test-account
      displayName: Zach Buchheit Test Account
  service-account-key:
    type: gcp:serviceAccount:Key
    # type: gcp:serviceaccount:Key
    properties:
      serviceAccountId: ${service-account.id}
    options:
      parent: ${service-account}
```
</details>

## Steps to reproduce
1. Run pulumi up with the program and pulumi plugin `6.67.1`
1. Install pulumi plugin `7.16.0`
1. Change the type from `serviceAccount` to `serviceaccount` in the code
1. Run `pulumi up --target "**bucket**"`
1. Witness error of missing parent resource

### Output of `pulumi about`

```
CLI          
Version      3.112.0
Go Version   go1.22.1
Go Compiler  gc

Plugins
NAME  VERSION
gcp   unknown
yaml  unknown

Host     
OS       darwin
Version  14.2.1
Arch     arm64

This project is written in yaml

Current Stack: zbuchheit-pulumi-corp/service-account/dev

TYPE                                URN
pulumi:pulumi:Stack                 urn:pulumi:dev::service-account::pulumi:pulumi:Stack::service-account-dev
pulumi:providers:gcp                urn:pulumi:dev::service-account::pulumi:providers:gcp::default
gcp:serviceAccount/account:Account  urn:pulumi:dev::service-account::gcp:serviceAccount/account:Account::service-account
gcp:storage/bucket:Bucket           urn:pulumi:dev::service-account::gcp:storage/bucket:Bucket::my-bucket-zbuchheit
gcp:serviceAccount/key:Key          urn:pulumi:dev::service-account::gcp:serviceAccount/account:Account$gcp:serviceAccount/key:Key::service-account-key


Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/zbuchheit-pulumi-corp
User           zbuchheit-pulumi-corp
Organizations  zbuchheit-pulumi-corp,
Token type     personal

No dependencies found

Pulumi locates its logs in /var/folders/lh/l71cdh810xb33t0jc7qmt5_80000gn/T/ by default
```

### Additional context

While a pulumi up will just work, I think it is a reasonable expectation of users to be able to leverage --target in a major upgrade to help feel confidence in the changes they are making.

Additionally, there is a workaround to just target the serviceaccount, which kind of strangely displays no diff, but will actually change the name in the state, then update the other resources, but this behavior does feel to be a bug in the engine of some sort and want to bring awareness to it early.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by zbuchheit** [comment]
good to know thank you @guineveresaenger !

**Comment by zbuchheit** [comment]
updated title to be more specific to GCP since we moved it into pulumi/pulumi

**Comment by lunaris** [comment]
Thanks for the great bug report @zbuchheit 🎉 So I've done some digging into this, and I believe it's already been fixed by #17117. The setup for the bug is as follows:

* Register a resource, `P`, of type `pkg:module:R1`, that will act as a parent
    - in @zbuchheit's example, this is the account
* Register a resource, `C`, of type `pkg:module:R1`, that will act as a child
    - in @zbuchheit's example, this is the account key
* Register a resource, `T`, of any type (`pkg:module:T`, say), that will act as a `--target`. `T` has no relationship with either `P` or `C`.
    - in @zbuchheit's example, this is the bucket

Run an operation that puts all these in the state. Now, make the following changes:

* Change the types of `P` and `C` to e.g. `pkg:module:R2`
    - this is analogous to the change from `serviceAccount` to `serviceaccount` in the GCP package between v6 and v7
* Add aliases to `P` and `C` referring to their old types (e.g. with `aliases=[pulumi.Alias(type_="pkg:module:R1")]` in Python)
    - internally, the v7 version of the GCP provider auto-adds aliases in this manner to preserve `serviceAccount` resources that are now `serviceaccount` resources
* Run an update that only `--target`s `T`

Prior to #17117, the fact that only `T` (which has no relationship to `P` or `C`) was targeted would mean that the changing/aliased relationship between `P` and `C` would not be preserved and the error mentioned in the report would be thrown by `checkParent` in `step_generator`. Since that has merged, this should no longer be the case. I've (I believe) confirmed this by:

* `git reset --hard v3.112.0` (the version in the report)
* Ran a lifecycle test with the above scenario and reproduced the error
* `git reset --hard origin/master` (`HEAD`)
* Ran the same test and observed no failure

_Note: the test I ran is a variant of `TestUntargetedDependencyChainsArePreserved`, which was committed as part of #17117 and (I believe) covers this case now and in future_

@zbuchheit, would you be able to try reproducing this on both v3.112.0 and e.g. v3.138.0 and confirm that it's no longer an issue? If you're not able I can also opportunistically close this. Thanks again for the detailed report and minimal reproduction 🙏

**Comment by zbuchheit** [comment]
I think it is reasonable to opportunistically close this. I might have some time tomorrow to review, but given your detailed analysis I think you have enough evidence that this has been fixed.

**Comment by lunaris** [comment]
Fixed by #17117
